### DISCUSSION TITLE
Throw TypeConverterException from custom TypeConverter

### DISCUSSION THREAD

**Discussion by ben-private** [JoshClose/CsvHelper#2282] [Q&A] [answered]
Hey im having a custom Converter who converts boolean values to localised strings:
```
public class VpBoolCsvConverter : TypeConverter<bool>
{
    public override bool ConvertFromString(string? text, IReaderRow row, MemberMapData memberMapData)
    {
        return text switch
        {
            "" or "Nein" => false,
            "Ja" => true,
            _ => throw new TypeConverterException(this, memberMapData, text, null),
        };
    }

    public override string? ConvertToString(bool value, IWriterRow row, MemberMapData memberMapData)
    {
        return value ? "Ja" : "Nein";
    }
}
```

If this conversion fails i want to throw a TypeConverterException so i can handle it along the other default conversion errors when reading using the ReadingExceptionOccurred Property of the CsvReader Configuration:
```
ReadingExceptionOccurred = (ReadingExceptionOccurredArgs args) =>
{
    var ex = args.Exception;
    if (ex is TypeConverterException typeConverterEx)
    {
        var rowNumber = typeConverterEx.Context!.Parser!.Row;
        var column = typeConverterEx.Context.Reader!.HeaderRecord?[typeConverterEx.Context.Reader.CurrentIndex];
        result.Messages.Add($"Error... {typeConverterEx.Text} .. {column} .. {rowNumber}.");
        return false;
    }

    throw ex;
}
```

My Issue now is that im missing the CsvContext in the TypeConverter but it is needed for the TypeConverterException, otherwise the exception handler falls apart. So my question is there someway to retrieve the CsvContext in the Converter or can i throw another exceptions which gets plugged into the TypeConverterException somewhere upstream? Any other ideas?

Thanks very much.

**Comment by AltruCoder** [comment, marked as answer]
You should be able to use `row.Context`
```csharp
_ => throw new TypeConverterException(this, memberMapData, text, row.Context),
```

Your other option would be to use `TypeConverterOptionsCache` instead of the `VpBoolCsvConverter`
```csharp
csv.Context.TypeConverterOptionsCache.GetOptions<bool>().BooleanTrueValues.Add("Ja");
csv.Context.TypeConverterOptionsCache.GetOptions<bool>().BooleanFalseValues.Add("Nein");
csv.Context.TypeConverterOptionsCache.GetOptions<bool>().BooleanFalseValues.Add("");
```

**Reply by ben-private** [reply]
Thanks so much, the TypeConverterOptionsCache works like a charme
